### ISSUE TITLE
Placement of "optional" on a property or method

### ISSUE THREAD

**Issue by andy-ms** [JamesHenry/tsep-babylon-test#13]
### Code

```ts
class C {
    m?: number
}
```

### Eslint

```json
"key": Object {
  "name": "m",
  "optional": true,
  "type": "Identifier",
},
```

### Babylon

```json
"key": Object {
  "name": "m",
  "type": "Identifier",
},
"optional": true,
```

**Comment by JamesHenry** [comment]
I would say that this is an ESLint issue. I think `optional` here is conceptually the same as the standard `computed` boolean, and that would live outside the key, as it does in the Babylon one.

Incidentally, if we amend the source slightly and think about `m` being computed:

```
class C {
    ['m']?: number
}
```

...`optional` is actually completely missing from the ESLint AST, but correctly set on the Babylon one.

**Comment by JamesHenry** [comment]
@andy-ms This is more of an awkward one that I first realised.

In Flow, it is the `Identifier` node which is used to host the `optional` flag, and that is what ESLint is doing.

E.g. on function parameters:

```ts
function foo(x?) {}
```
```
"params": [
        {
          "type": "Identifier",
          "loc": {
            "source": null,
            "start": {
              "line": 5,
              "column": 13
            },
            "end": {
              "line": 5,
              "column": 15
            }
          },
          "range": [
            42,
            44
          ],
          "name": "x",
          "typeAnnotation": null,
          "optional": true
        }
      ],
```

@andy-ms I am assuming you have matched this structure for your parsing of optional function parameters?

This then is also the case in ESLint for `ClassProperty`s as you have noted above, however things get confusing when we consider the computed property case that I demonstrated in my last comment, because there is no longer an `Identifier` node at all...

See also https://github.com/eslint/typescript-eslint-parser/pull/292

cc @soda0289

**Comment by andy-ms** [comment]
Yes, babylon is adding `optional` to the Identifier in the case of a parameter since there's nothing else to add it to.
